Date-changing command +cheat interface (#4184)

hgentry opened the pull request:
A command line function to change the date as YYYY MM DD, YYYY MM, or YYYY.

Ex:

`$ date 0005 10 15` will set the date to October 15th, Year 5. Omitting month or day will preserve the existing values.

There's also an interface in the Cheats menu for setting the date manually.

Edit: to be me more specific, it accepts [YYY]Y [[M]M [[D]D]], but that's a bit of a mouthful.

Gymnasiast commented on src/openrct2/localisation/date.h:
```diff
@@ -61,14 +61,16 @@ extern openrct_timeofday gRealTimeOfDay;
 bool date_is_fortnight_start(sint32 monthTicks);
 bool date_is_month_start(sint32 monthTicks);
-
+static const sint16 DaysInMonth[MONTH_COUNT] = { 31, 30, 31, 30, 31, 31, 30, 31 };
```
I'm sure these must have been defined somewhere.

AaronVanGeffen commented on src/openrct2/localisation/string_ids.h:
```diff
@@ -3748,105 +3748,112 @@ enum {
-    STR_COPY_SELECTED = 6129,
-    STR_COPY_ALL = 6130,
-    STR_OBJECT_SOURCE = 6131,
+STR_OBJECT_TYPE = 6125,
```
You've accidentally removed the indentation here. Could you please re-introduce it?

hgentry commented:
Latest commit addresses comments.

AaronVanGeffen commented on src/openrct2/localisation/date.c:
```diff
@@ -67,6 +67,71 @@ void date_reset()
+    if (argc == 1)
+    {
+        year = atoi(argv[0]);
+        if (year < 1 || year > 8192) {
```
There is quite a bit of overlap in each of these three cases. I think this could be improved upon by moving the common parts, e.g. getting the year and setting the date, outside of the if-blocks.

AaronVanGeffen commented on src/openrct2/Cheats.h:
```diff
@@ -95,6 +95,7 @@ enum {
     CHEAT_DISABLERIDEVALUEAGING,
     CHEAT_IGNORERESEARCHSTATUS,
     CHEAT_ENABLEALLDRAWABLETRACKPIECES,
+	CHEAT_DATE_SET,
```
This is very minor, but you've used a tab character instead of four spaces.

AaronVanGeffen commented:
Thanks for taking the time to address the formatting issues. I've left another comment with regards to code duplication in your `cmdline_for_date` function. Other than that, it looks fine and just needs some testing.

hgentry commented:
Sounds great. Updates are in.

AaronVanGeffen commented on src/openrct2-ui/windows/Cheats.cpp:
```diff
@@ -940,9 +1000,34 @@ static void window_cheats_paint(rct_window *w, rct_drawpixelinfo *dpi)
     else if(w->page == WINDOW_CHEATS_PAGE_MISC){
+        sint32 monthstr = 0;
+        switch (month_spinner_value) {
+        case 1: monthstr = STR_MONTH_MARCH;
```
This is basically `DateGameMonthNames` in `localisation.c`. It's a bit of a waste to duplicate it here…

You can leverage it by making a new string `STR_FORMAT_MONTH` that uses the `{MONTH}` formatting code to use in your call to `gfx_draw_string_right` below. The argument can be passed using `set_format_arg(offset, type, value)`.

AaronVanGeffen commented:
I've taken it out for a quick spin. It works well, but I have a few suggestions:

* The cheat window is getting really long with the date settings in the third tab, intended for park settings. I understand it's not ideal, but could you move it into the first tab (money), for now? We'll have to reorganise the cheats window at some point in the near future, either way. (It's on my list…)

* The bottom-right toolbar isn't invalidated when the date is set or reset. Please use the `window_invalidate_by_class` function to do this.

Other than that, it seems to be working well.

hgentry commented:
Changes made.

AaronVanGeffen commented:
Cheers, that looks much better. The bottom toolbar is addressed, too.

There seems to be an issue repainting the widgets in the cheat window, though. If I change either day, month or year through the spinners, the corresponding widget isn't properly redrawn. I'm guessing you may need to invalidate them, still?

One final note: please don't merge `develop` back into your branch. Rather, [rebase your branch](https://help.github.com/articles/about-git-rebase/) on top of develop. This makes the history more consistent. You can also use it to squash commits together, incidentally, which I would recommend you do, too.

Thanks again — this is shaping out quite nicely.

rwjuk commented:
With regard to rebasing, [here's](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) a good guide to that if you need it.

hgentry commented:
Hold on a bit... I tried to rebase and broke everything

AaronVanGeffen commented:
When rebasing, remember to use `push -f` to push your commits to GitHub.

Gymnasiast commented on src/openrct2-ui/windows/Cheats.cpp:
```diff
@@ -937,12 +998,19 @@ static void window_cheats_paint(rct_window *w, rct_drawpixelinfo *dpi)
     else if(w->page == WINDOW_CHEATS_PAGE_MISC){
         gfx_draw_string_left(dpi, STR_CHEAT_STAFF_SPEED,            nullptr,   COLOUR_BLACK, w->x + XPL(0) + TXTO, w->y + YPL(17) + TXTO);
         gfx_draw_string_left(dpi, STR_FORCE_WEATHER,                nullptr,   COLOUR_BLACK, w->x + XPL(0) + TXTO, w->y + YPL(10) + TXTO);
-        gfx_draw_string_right(dpi, STR_FORMAT_INTEGER,      &park_rating_spinner_value, w->colours[1], w->x + WPL(1) - 10 - TXTO, w->y + YPL(5) + TXTO);
```
This is not related to your PR, why did you move it?

hgentry replied:
Didn't mean to. Cleanup of the rebase is still in progress

Gymnasiast commented on src/openrct2-ui/windows/Cheats.cpp:
```diff
@@ -552,7 +613,7 @@ static void window_cheats_misc_mousedown(rct_window *w, rct_widgetindex widgetIn
         dropdown_set_checked(currentWeather, true);
     }
-    break;
+        break;
```
Should be reverted.

Gymnasiast commented:
I just rebased your branch. Now you overwrote it with a non-mergeable version again.

AaronVanGeffen commented:
@hgentry It looks like you're rebasing on top of an outdated local develop branch. Please make sure it's up to date before you rebase.

hgentry commented:
Sure, go ahead. I didn't know you were doing that; I was fixing it myself at the same time.

hgentry commented:
Out of  curiosity, why was it being marked unmergeable? There were no conflicts on the change list at any point.

AaronVanGeffen commented:
> There were no conflicts on the change list at any point.

There were, when https://github.com/OpenRCT2/OpenRCT2/pull/6944 was merged.

hgentry commented:
Must've missed it, then. Anyway, I still need to add changes for the last two comments and then it'll be good to go.

Gymnasiast commented:
There were. Mostly in the includes list. The thing is that if you edit line 19 and upstream changed line 18, it will create a merge conflict - they're too close together.

You probably tried to rebase against your own develop or against an outdated version of our develop. Always make sure you have the newest develop fetched.

Also, you also need to do this now to keep my changes:
```
git fetch origin
git reset --hard origin/develop2
```

hgentry commented:
Should I squash that last commit, or are we good to go?

hgentry commented:
Hold on a moment- something weird happened at some point and there is a whole lot of stuff where it shouldn't be.

AaronVanGeffen commented on src/openrct2/localisation/Date.cpp:
```diff
@@ -65,6 +66,67 @@ void date_reset()
+
+void date_set(sint32 year, sint32 month, sint32 day)
+{
+    year = clamp(1, year, 8192);
```
With `date.c` renamed to `Date.cpp`, and now being compiled as C++, these should be changed to read `Math::Clamp` rather than `clamp`. Sorry about that.

hgentry replied:
Changed

hgentry commented:
Should I squash those last few commits?

AaronVanGeffen commented:
Ideally, yes, but that can be done at a later point. First, I believe @Gymnasiast is eager to take it for a test run, too. 😉 

Thanks for sticking with it! Hopefully, it has been a nice introduction to the code base, so far.

AaronVanGeffen commented:
Widgets weren't properly invalidated when using the spinner buttons. I've pushed a fix for this.

AaronVanGeffen commented:
@hgentry I've rebased your PR after the merge of #6956 caused some new conflicts, and also moved the console functions to the appropriate file.

It looks like you're not in the `contributors.md` file yet. If you like, you can add yourself there. I think this change also qualifies to be listed in the `changelog.txt` file, so if you like, you can add a line there, too.

janisozaur commented:
https://github.com/OpenRCT2/OpenRCT2/issues/4184

hgentry commented:
> I've rebased your PR after the merge of #6956 caused some new conflicts, and also moved the console functions to the appropriate file.

Sounds good, some of your commit was stuff I lost when I messed up the rebase and the function being in the wrong place is from being new to the codebase :/. Sorry about the need for you to do changes and to clean up after me; all my git experience (including professional) has ignored rebase and just pulled the dev branch before PR.

> It looks like you're not in the contributors.md file yet. If you like, you can add yourself there. I think this change also qualifies to be listed in the changelog.txt file, so if you like, you can add a line there, too.

Sure. I'll take care of that when I get home in a bit.

AaronVanGeffen commented:
No worries, we all have to learn these things some time. Happy to see you picked up something you'd like to do, and that you made a PR for it, really! You're doing fine.

Gymnasiast commented:
I also have to add that the amount of changes to the code base these days has been enormous. We have had to rebase every PR multiple times. Also, we are more than willing to help people who want to contribute and are eager to learn something new.

AaronVanGeffen commented on src/openrct2/localisation/Date.h:
```diff
@@ -61,8 +61,10 @@ extern openrct_timeofday gRealTimeOfDay;
 sint32 date_get_month(sint32 months);
 sint32 date_get_year(sint32 months);
 sint32 date_get_total_months(sint32 month, sint32 year);
+sint32 cmdline_for_date(const utf8 **argv, sint32 argc);
```
Note to self: forgot to remove this line. Should probably mark `cmdline_for_date` as static in Console.cpp, too.

hgentry commented:
Thanks for the kind words! I've been enjoying working on this and look forward to doing more.
